src/bin/tail.rs

```diff
@@ -0,0 +1,185 @@
+                        }
+                    }
+                    Err(err) => return Err(err),
+                };
+            }
+
+            for mut line in deque {
+                line.push('\n');
+                try!(writer.write_all(line.as_bytes()));
+            }
+        }
+    } else {
+        if opts.skip {
+            let bytes = input.bytes().skip(opts.num);
+
+            for byte_res in bytes {
+                match byte_res {
+                    Ok(byte) => try!(writer.write_all(&[byte])),
+                    Err(err) => return Err(err),
+                };
+            }
+        } else {
+            let bytes = input.bytes();
+            let mut deque = VecDeque::new();
+
+            for byte_res in bytes {
+                match byte_res {
+                    Ok(byte) => {
+                        deque.push_back(byte);
+
+                        if deque.len() > opts.num {
+                            deque.pop_front();
+                        }
+                    }
+                    Err(err) => return Err(err),
+                };
+            }
+
+            for byte in deque {
+                try!(writer.write_all(&[byte]));
+            }
+        }
+    }
+
+    Ok(())
+}
+
+fn main() {
+    // default options
+    let mut opts = Options {
+        lines: true,
+        num: 10,
+        skip: false,
+    };
+    let stdout = io::stdout();
+    let mut stdout = stdout.lock();
+    let mut stderr = io::stderr();
+    let mut args = env::args().skip(1);
+    let mut paths: Vec<String> = Vec::new();
+
+    // parse options
+    while let Some(arg) = args.next() {
+        if arg.starts_with('-') {
+            match arg.as_str() {
+                "-h" | "--help" => {
+                    stdout.write_all(MAN_PAGE.as_bytes()).try(&mut stderr);
+                    return;
+                }
+                "-n" | "--lines" => opts.lines = true,
+                "-c" | "--bytes" => opts.lines = false,
+                _ => fail("invalid option", &mut stderr),
+            }
+
+            if let Some(arg) = args.next() {
+                if arg.starts_with('+') {
+                    opts.skip = true;
+                }
+
+                opts.num = arg.parse::<usize>().try(&mut stderr);
+            } else {
+                fail("missing argument (number of lines/bytes)", &mut stderr);
+            }
+        } else {
+            paths.push(arg);
+        }
+    }
+
+    // run the main part
+    if paths.is_empty() {
+        let stdin = io::stdin();
+        let stdin = stdin.lock();
+        tail(stdin, stdout, opts).try(&mut stderr);
+    } else if paths.len() == 1 {
+        let file = fs::File::open(&paths[0]).try(&mut stderr);
+        tail(file, stdout, opts).try(&mut stderr);
+    } else {
+        for path in paths {
+            let file = fs::File::open(&path).try(&mut stderr);
+            writeln!(&mut stdout, "==> {} <==", path).try(&mut stderr);
```

Comment:
This is slow, since it does another heap allocation. Instead do:

```rust
stdout.write(b"==> ").try(&mut stderr);
stdout.write(path.as_bytes()).try(&mut stderr);
stdout.writeln(b"<==").try(&mut stderr);
```